### PR TITLE
Prepare generators rename

### PR DESCRIPTION
**Pull request by bigfoot90**
## Subject

I think these names are more consistent with current implementations as they are not related to kind of persistence by to the type ID used by the model.

I am targeting this branch, it is supposed to be BC.

Reply to https://github.com/sonata-project/SonataMediaBundle/issues/1342#issuecomment-581243831

## Changelog

```markdown
### Changed
- Renamed class `Sonata\MediaBundle\Command\DefaultGenerator` into `Sonata\MediaBundle\Command\IdGenerator`
- Renamed class `Sonata\MediaBundle\Command\ODMGenerator` into `Sonata\MediaBundle\Command\UuidGenerator`
- Renamed class `Sonata\MediaBundle\Command\PHPCRGenerator` into `Sonata\MediaBundle\Command\PathGenerator`
```

### CODE REVIEW

**Comment by bigfoot90** [comment]
Sorry, I don't undestand why these commits 2f0e3c3 and d6b3806 are here in this PR, they should be already merged.

**Comment by core23** [comment]
Can you do this on the stable branch? Introduce the new classes and deprecate the old one. This will improve the upgrade path to the next major release.

**Comment by bigfoot90** [comment]
Rebased

**Comment by bigfoot90** [comment]
Rebased

File: src/Generator/IdGenerator.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+namespace Sonata\MediaBundle\Generator;
+
+use Sonata\MediaBundle\Model\MediaInterface;
+
+/**
+ * @final since sonata-project/media-bundle 3.21.0
+ */
+class IdGenerator implements GeneratorInterface
```
Comment (greg0ire) [inline src/Generator/IdGenerator.php]: You can make it final rightaway, it's not a BC-break

File: src/Generator/IdGenerator.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * @var int
+     */
+    protected $secondLevel;
+
+    /**
+     * @param int $firstLevel
+     * @param int $secondLevel
+     */
+    public function __construct($firstLevel = 100000, $secondLevel = 1000)
```
Comment (greg0ire) [inline src/Generator/IdGenerator.php]: You can use type declarations rightaway

Comment (OskarStark) [reply src/Generator/IdGenerator.php]: And remove the phpdoc as it does not add any value

File: src/Generator/IdGenerator.php (hunk 1 above)
Comment (bigfoot90) [reply src/Generator/IdGenerator.php]: It's actually extended by the deprecated `DefaultGenerator` right now.
https://github.com/sonata-project/SonataMediaBundle/pull/1688/files#diff-fcd815cd554f9f3cf32406cc1f5b1831R21

Comment (core23) [reply src/Generator/IdGenerator.php]: You can still make this final  by proxing the old methods

Comment (greg0ire) [reply src/Generator/IdGenerator.php]: Or better, use a `class_alias`, like in this PR: https://github.com/doctrine/persistence/pull/71

**Comment by OskarStark** [comment]
PR Header and title should be updated, as its no breaking change anymore!

File: src/Generator/IdGenerator.php (hunk 1 above)
Comment (greg0ire) [reply src/Generator/IdGenerator.php]: @bigfoot90 don't hesitate to ask me if you want some assistance / an example for this.

File: src/Generator/IdGenerator.php (hunk 2 above)
Comment (bigfoot90) [reply src/Generator/IdGenerator.php]: But then also `DefaultGenerator` class will be using typed declarations.
Isn't this a BC break?

Comment (greg0ire) [reply src/Generator/IdGenerator.php]: Ah you are correct, if we use a class aliases, you can't do this. I hadn't that in mind when writing this comment.

Comment (greg0ire) [reply src/Generator/IdGenerator.php]: even if you use extension, in fact.

**Comment by greg0ire** [comment]
The build is failing because there are self deprecation notices. This can happen either because we are testing the legacy paths without marking corresponding tests with the `@group legacy` annotation, or because there are still places in `src` where we call the legacy paths.

File: tests/Generator/ODMGeneratorTest.php [hunk 3]
```diff
@@ -17,6 +17,9 @@
 use Sonata\MediaBundle\Generator\ODMGenerator;
 use Sonata\MediaBundle\Tests\Entity\Media;
 
+/**
+ * @deprecated
```
Comment (greg0ire) [inline tests/Generator/ODMGeneratorTest.php]: a test is public API, it cannot be deprecated. Let's mark it as covering a legacy path.
```suggestion
 * @group legacy
```

File: tests/Generator/PHPCRGeneratorTest.php [hunk 4]
```diff
@@ -17,6 +17,9 @@
 use Sonata\MediaBundle\Generator\PHPCRGenerator;
 use Sonata\MediaBundle\Tests\Entity\Media;
 
+/**
+ * @deprecated
```
Comment (greg0ire) [inline tests/Generator/PHPCRGeneratorTest.php]:
```suggestion
 * @group legacy
```

File: tests/Generator/DefaultGeneratorTest.php [hunk 5]
```diff
@@ -17,6 +17,9 @@
 use Sonata\MediaBundle\Generator\DefaultGenerator;
 use Sonata\MediaBundle\Tests\Entity\Media;
 
+/**
+ * @deprecated
```
Comment (greg0ire) [inline tests/Generator/DefaultGeneratorTest.php]:
```suggestion
 * @group legacy
```

**Comment by greg0ire** [comment]
The build is failing with only one deprecation now:  1x: The Sonata\MediaBundle\Generator\DefaultGenerator class is deprecated since sonata/media-bundle 3.4 and will be removed in 4.0. Use \Sonata\MediaBundle\Generator\IdGenerator instead.

    1x in MediaControllerTest::testGetMediaAction from Sonata\MediaBundle\Tests\Functional

It means something in that test probably needs to be changed so that the `DefaultGenerator` is no longer called.

**Comment by greg0ire** [comment]
I squashed into one commit, and added more changes (some tests needed to switch to `IdGenerator`). I think the build will pass. One thing left to do: add an upgrade note in `UPGRADE-3.x.md`. Please do so, amend and force push.

**Comment by greg0ire** [comment]
It passes! Don't worry about the upgrade note, I'm taking care of it for you.

File: UPGRADE-3.x.md [hunk 6]
```diff
@@ -1,6 +1,18 @@
...
 
+UPGRADE FROM 3.23 to 3.24
+=========================
+
+### Generators rename
+
+Some classes have been renamed, the old types are just aliases for the new ones.
+You should replace all references to them in your code.
+
+- `Sonata\MediaBundle\Command\DefaultGenerator` is deprecated in favor of `Sonata\MediaBundle\Command\IdGenerator`
```
Comment (core23) [inline UPGRADE-3.x.md]:
```suggestion
- `Sonata\MediaBundle\Command\DefaultGenerator` is deprecated in favor of `Sonata\MediaBundle\Generator\IdGenerator`
```

Same for the other

File: src/Generator/IdGenerator.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $this->firstLevel = $firstLevel;
+        $this->secondLevel = $secondLevel;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (core23) [inline src/Generator/IdGenerator.php]: Please remove this useless comment

File: src/Generator/PathGenerator.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+use Sonata\MediaBundle\Model\MediaInterface;
+
+/**
+ * @final since sonata-project/media-bundle 3.21.0
+ */
+class PathGenerator implements GeneratorInterface
+{
+    /**
+     * {@inheritdoc}
```
Comment (core23) [inline src/Generator/PathGenerator.php]: Please remove this comment

File: src/Generator/UuidGenerator.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+namespace Sonata\MediaBundle\Generator;
+
+use Sonata\MediaBundle\Model\MediaInterface;
+
+class UuidGenerator implements GeneratorInterface
+{
+    /**
+     * {@inheritdoc}
```
Comment (core23) [inline src/Generator/UuidGenerator.php]: ... and this

File: tests/Generator/IdGeneratorTest.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\MediaBundle\Tests\Generator;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\MediaBundle\Generator\IdGenerator;
+use Sonata\MediaBundle\Tests\Entity\Media;
+
+class IdGeneratorTest extends TestCase
```
Comment (core23) [inline tests/Generator/IdGeneratorTest.php]:
```suggestion
final class IdGeneratorTest extends TestCase
```

File: tests/Generator/PathGeneratorTest.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\MediaBundle\Tests\Generator;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\MediaBundle\Generator\PathGenerator;
+use Sonata\MediaBundle\Tests\Entity\Media;
+
+class PathGeneratorTest extends TestCase
```
Comment (core23) [inline tests/Generator/PathGeneratorTest.php]:
```suggestion
final class PathGeneratorTest extends TestCase
```

File: tests/Generator/UuidGeneratorTest.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\MediaBundle\Tests\Generator;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\MediaBundle\Generator\UuidGenerator;
+use Sonata\MediaBundle\Tests\Entity\Media;
+
+class UuidGeneratorTest extends TestCase
```
Comment (core23) [inline tests/Generator/UuidGeneratorTest.php]:
```suggestion
final class UuidGeneratorTest extends TestCase
```

**Comment by greg0ire** [comment]
Thanks @bigfoot90 !

**Pull request merged**